File: backend/app/api/v1/sharing.py
import secrets
import uuid

from fastapi import APIRouter, Depends
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.database import get_db
from app.dependencies import get_current_user
from app.models.user import User
from app.models.scoring import ScoringSession
from app.schemas.scoring import ShareLinkOut, SharedSessionOut, EndOut
from app.core.exceptions import NotFoundError, AuthError
from app.config import settings

router = APIRouter(prefix="/share", tags=["sharing"])


@router.post("/sessions/{session_id}", response_model=ShareLinkOut)
async def create_share_link(
    session_id: uuid.UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(ScoringSession).where(
            ScoringSession.id == session_id,
            ScoringSession.user_id == user.id,
        )
    )
    session = result.scalar_one_or_none()
    if not session:
        raise NotFoundError("Session not found")

    if not session.share_token:
        session.share_token = secrets.token_urlsafe(16)
        await db.commit()
        await db.refresh(session)

    return ShareLinkOut(
        share_token=session.share_token,
        url=f"{settings.FRONTEND_URL}/shared/{session.share_token}",
    )


@router.delete("/sessions/{session_id}")
async def revoke_share_link(
    session_id: uuid.UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(ScoringSession).where(
            ScoringSession.id == session_id,
            ScoringSession.user_id == user.id,
        )
    )
    session = result.scalar_one_or_none()
    if not session:
        raise NotFoundError("Session not found")

    session.share_token = None
    await db.commit()
    return {"detail": "Share link revoked"}


@router.get("/s/{token}", response_model=SharedSessionOut)
async def get_shared_session(
    token: str,
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(ScoringSession)
        .where(ScoringSession.share_token == token)
        .options(selectinload(ScoringSession.user))
    )
    session = result.scalar_one_or_none()
    if not session:
        raise NotFoundError("Shared session not found")

    return SharedSessionOut(
        archer_name=session.user.display_name or session.user.username,
        archer_avatar=session.user.avatar,
        template_name=session.template.name if session.template else "Unknown",
        template=session.template,
        total_score=session.total_score,
        total_x_count=session.total_x_count,
        total_arrows=session.total_arrows,
        notes=session.notes,
        location=session.location,
        weather=session.weather,
        started_at=session.started_at,
        completed_at=session.completed_at,
        ends=[EndOut.model_validate(e) for e in session.ends],
    )

File: backend-go/internal/handler/sharing.go
package handler

import (
	"context"
	"crypto/rand"
	"encoding/base64"
	"net/http"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"

	"github.com/quiverscore/backend-go/internal/config"
	"github.com/quiverscore/backend-go/internal/middleware"
	"github.com/quiverscore/backend-go/internal/repository"
)

type SharingScoringRepository interface {
	GetShareToken(ctx context.Context, sessionID, userID string) (*string, error)
	SetShareToken(ctx context.Context, sessionID, token string) error
	RevokeShareToken(ctx context.Context, sessionID, userID string) (bool, error)
	GetSharedSession(ctx context.Context, token string) (*repository.SharedSessionData, error)
	LoadEnds(ctx context.Context, sessionID string) ([]repository.EndOut, error)
}

type SharingUserRepository interface {
	GetArcherInfo(ctx context.Context, userID string) (username string, displayName, avatar *string, err error)
}

type SharingRoundRepository interface {
	Get(ctx context.Context, id string) (*repository.RoundTemplateOut, error)
}

type SharingHandler struct {
	Scoring SharingScoringRepository
	Users   SharingUserRepository
	Rounds  SharingRoundRepository
	Cfg     *config.Config
}

func (h *SharingHandler) Routes(r chi.Router) {
	r.With(middleware.RequireAuth(h.Cfg.SecretKey)).Post("/sessions/{id}", h.CreateShareLink)
	r.With(middleware.RequireAuth(h.Cfg.SecretKey)).Delete("/sessions/{id}", h.RevokeShareLink)
	r.Get("/s/{token}", h.GetSharedSession)
}

// ── Types ─────────────────────────────────────────────────────────────

type shareLinkOut struct {
	ShareToken string `json:"share_token"`
	URL        string `json:"url"`
}

type sharedSessionOut struct {
	ArcherName   string                   `json:"archer_name"`
	ArcherAvatar *string                  `json:"archer_avatar"`
	TemplateName string                   `json:"template_name"`
	Template     *repository.RoundTemplateOut `json:"template"`
	TotalScore   int                      `json:"total_score"`
	TotalXCount  int                      `json:"total_x_count"`
	TotalArrows  int                      `json:"total_arrows"`
	Notes        *string                  `json:"notes"`
	Location     *string                  `json:"location"`
	Weather      *string                  `json:"weather"`
	StartedAt    time.Time                `json:"started_at"`
	CompletedAt  *time.Time               `json:"completed_at"`
	Ends         []repository.EndOut      `json:"ends"`
}

// ── Create Share Link ─────────────────────────────────────────────────

func (h *SharingHandler) CreateShareLink(w http.ResponseWriter, r *http.Request) {
	sessionID := chi.URLParam(r, "id")
	if _, err := uuid.Parse(sessionID); err != nil {
		Error(w, http.StatusNotFound, "Session not found")
		return
	}

	userID := middleware.GetUserID(r.Context())
	ctx := r.Context()

	shareToken, err := h.Scoring.GetShareToken(ctx, sessionID, userID)
	if err != nil {
		Error(w, http.StatusNotFound, "Session not found")
		return
	}

	if shareToken == nil || *shareToken == "" {
		token := generateURLSafeToken(16)
		if err := h.Scoring.SetShareToken(ctx, sessionID, token); err != nil {
			Error(w, http.StatusInternalServerError, "Internal server error")
			return
		}
		shareToken = &token
	}

	JSON(w, http.StatusOK, shareLinkOut{
		ShareToken: *shareToken,
		URL:        h.Cfg.FrontendURL + "/shared/" + *shareToken,
	})
}

// ── Revoke Share Link ─────────────────────────────────────────────────

func (h *SharingHandler) RevokeShareLink(w http.ResponseWriter, r *http.Request) {
	sessionID := chi.URLParam(r, "id")
	if _, err := uuid.Parse(sessionID); err != nil {
		Error(w, http.StatusNotFound, "Session not found")
		return
	}

	userID := middleware.GetUserID(r.Context())

	ok, err := h.Scoring.RevokeShareToken(r.Context(), sessionID, userID)
	if err != nil || !ok {
		Error(w, http.StatusNotFound, "Session not found")
		return
	}

	JSON(w, http.StatusOK, map[string]string{"detail": "Share link revoked"})
}

// ── Get Shared Session ────────────────────────────────────────────────

func (h *SharingHandler) GetSharedSession(w http.ResponseWriter, r *http.Request) {
	token := chi.URLParam(r, "token")
	ctx := r.Context()

	data, err := h.Scoring.GetSharedSession(ctx, token)
	if err != nil {
		Error(w, http.StatusNotFound, "Shared session not found")
		return
	}

	// Load archer info
	username, displayName, avatar, _ := h.Users.GetArcherInfo(ctx, data.UserID)
	archerName := username
	if displayName != nil && *displayName != "" {
		archerName = *displayName
	}

	// Load template
	var templateName string
	var template *repository.RoundTemplateOut
	t, err := h.Rounds.Get(ctx, data.TemplateID)
	if err == nil {
		templateName = t.Name
		template = t
	} else {
		templateName = "Unknown"
	}

	// Load ends
	ends, _ := h.Scoring.LoadEnds(ctx, data.SessionID)

	JSON(w, http.StatusOK, sharedSessionOut{
		ArcherName:   archerName,
		ArcherAvatar: avatar,
		TemplateName: templateName,
		Template:     template,
		TotalScore:   data.TotalScore,
		TotalXCount:  data.TotalXCount,
		TotalArrows:  data.TotalArrows,
		Notes:        data.Notes,
		Location:     data.Location,
		Weather:      data.Weather,
		StartedAt:    data.StartedAt,
		CompletedAt:  data.CompletedAt,
		Ends:         ends,
	})
}

// ── Helpers ───────────────────────────────────────────────────────────

func generateURLSafeToken(nBytes int) string {
	b := make([]byte, nBytes)
	rand.Read(b)
	return base64.URLEncoding.WithPadding(base64.NoPadding).EncodeToString(b)
}
